src/gfx/convert.cpp

```diff
@@ -0,0 +1,866 @@
+				png_uint_32 yStep = 1u << PNG_PASS_ROW_SHIFT(pass);
+
+				for (png_uint_32 y = PNG_PASS_START_ROW(pass); y < height; y += yStep) {
+					png_bytep ptr = row.data();
+
+					png_read_row(png, ptr, nullptr);
+					for (png_uint_32 x = PNG_PASS_START_COL(pass); x < width; x += xStep) {
+						Rgba rgba(ptr[0], ptr[1], ptr[2], ptr[3]);
+
+						colors.registerColor(rgba);
+						pixel(x, y) = rgba.cgbColor();
+						isGrayOnly &= rgba.isGray();
+						ptr += 4;
+					}
+				}
+			}
+		}
+
+		png_read_end(png,
+		             nullptr); // We don't care about chunks after the image data (comments, etc.)
+	}
+
+	~Png() { png_destroy_read_struct(&png, &info, nullptr); }
+
+	class TilesVisitor {
+		Png const &_png;
+		bool const _columnMajor;
+		uint32_t const _width, _height;
+		uint32_t const _limit = _columnMajor ? _height : _width;
+
+	public:
+		TilesVisitor(Png const &png, bool columnMajor, uint32_t width, uint32_t height)
+			: _png(png), _columnMajor(columnMajor), _width(width), _height(height) {}
+
+		class Tile {
+			Png const &_png;
+			uint32_t const _x, _y;
+
+		public:
+			Tile(Png const &png, uint32_t x, uint32_t y) : _png(png), _x(x), _y(y) {}
+
+			uint16_t pixel(uint32_t xOfs, uint32_t yOfs) const {
+				return _png.pixel(_x + xOfs, _y + yOfs);
+			}
+		};
+
+	private:
+		struct iterator {
+			TilesVisitor const &parent;
+			uint32_t const limit;
+			uint32_t x, y;
+
+		public:
+			std::pair<uint32_t, uint32_t> coords() const { return {x, y}; }
+			Tile operator*() const { return {parent._png, x, y}; }
+
+			iterator &operator++() {
+				auto [major, minor] = parent._columnMajor ? std::tie(y, x) : std::tie(x, y);
+				major += 8;
+				if (major == limit) {
+					minor += 8;
+					major = 0;
+				}
+
+				return *this;
+			}
+
+			bool operator!=(iterator const &rhs) const {
+				return coords() != rhs.coords(); // Compare the returned coord pairs
+			}
+		};
+
+	public:
+		iterator begin() const { return {*this, _width, 0, 0}; }
+		iterator end() const {
+			iterator it{*this, _width, _width - 8, _height - 8}; // Last valid one
+			return ++it; // Now one-past-last
+		}
+	};
+public:
+	TilesVisitor visitAsTiles(bool columnMajor) const {
+		return {*this, columnMajor, width, height};
+	}
+};
+
+class RawTiles {
+public:
+	/**
+	 * A tile which only contains indices into the image's global palette
+	 */
+	class RawTile {
+		std::array<std::array<size_t, 8>, 8> _pixelIndices{};
+
+	public:
+		// Not super clean, but it's closer to matrix notation
+		size_t &operator()(size_t x, size_t y) { return _pixelIndices[y][x]; }
+	};
+
+private:
+	std::vector<RawTile> _tiles;
+
+public:
+	/**
+	 * Creates a new raw tile, and returns a reference to it so it can be filled in
+	 */
+	RawTile &newTile() {
+		_tiles.emplace_back();
+		return _tiles.back();
+	}
+};
+
+struct AttrmapEntry {
+	size_t protoPaletteID;
+	uint16_t tileID : 10;
+	uint16_t unused : 3;
+	uint16_t xFlip  : 1;
+	uint16_t yFlip  : 1;
+	// 1 more unused bit
```

Comment:
It's [implementation-defined](https://en.cppreference.com/w/cpp/language/bit_field) whether these will all be packed into two bytes anyway (let alone whether they're packed with the unused bit last or first, lol).

Reply:
(So if the `unused` 3 bits here are to make this align with the superfamiconv GBA format of how bits are packed, don't bother, we can't assume anything about that.)

Reply:
It's so that these two bits are (hopefully) in the right place. Using two `bool`s wouldn't be useful.

Reply:
They're not guaranteed to be at any particular spot, or even in the same byte as each other.

Reply:
At least there is a chance they are, unlike with two `bool`s.

Reply:
I mean, nothing in the code should be depending on layout that's only left up to chance; so I don't see why the layout is worth trying for in the first place. (Sure, maybe most compilers will pack them as expected, but even if they do I wouldn't bet on whether the unused bit is 0th or 15th.)

Reply:
If your main concern is just keeping `sizeof struct AttrmapEntry` small, you can just leave out the `unused` field, and hope the other bits get packed into one `uint16_t`. But even then it doesn't matter, since `size_t`'s alignment makes it [16 bytes](https://godbolt.org/z/qMah64qE7) even if you use `bool`s.

Reply:
The code doesn't *depend* on the layout: no UB is being invoked here, so it will work correctly regardless of how the implementation decides to lay the `struct` out. ⇒ There is nothing to lose.

However, laying out the bits like this gives an opportunity for them to be laid out in a position that eliminates bit shifting from attrmap output, leaving only masking.

There is nothing to lose, but potential gain; and the change is unobtrusive. Then, why not do it?

Reply:
I didn't realize that was the reasoning behind it, since it's not clear to me that this *does* line up with GBC attrmap encoding. Tile IDs aren't 10 bits; they're 8 plus bit 3 of the second byte (`OAMF_BANK1`).

Anyway, this is another instance of optimization way before it's time to do so. Do we have any reason to think that CPU usage during  attrmap output will be a bottleneck where it's worth avoiding some shift and mask instructions? (Even if the attrmap were written to disk a thousand times after being created once, the disk-writing is still what I'd expect as the limiting factor.)

I'd personally write the x and y flip properties as `bool`s because that's what they are, on-or-off boolean flags.

Reply:
They are 10-bit because they are *technically* 9-bit plus the bank; this is because each bank contains 384 tiles.

A one-bit field is equivalent to a boolean.

Reply:
A one-bit field is a distractingly more complex alternative to a `bool`. Seeing it makes me think, did the writer expect these to be packed together? Does the code assume so anywhere? Is it directly memcpy'ing structs to/from files? Did the writer expect the `// 1 more unused bit` to be where that comment is? How closely were they reading the C++ bit field spec, and how well do I need to understand it to trust this code?

I just read over convert.cpp, and the one line this even has a chance of optimizing is `uint8_t attr = entry.xFlip << 5 | entry.yFlip << 6;`. Compare how [gcc and clang](https://godbolt.org/z/fvKoW8zr1) do; IMO this does not at all look worth having more complex code.

Reply:
[This](https://godbolt.org/z/qq69jbEYj) is more representative (unused bits in the right place, parameter not passed by copy).

Reply:
Still not worth a more complex struct: you save one `mov`/`movzx` and two `shl`/`sal`s per tile attribute, if the compiler does the (admittedly common) desired packing.